DQM/HLTEvF/plugins/TrigObjTnPSource.cc

```diff
@@ -0,0 +1,94 @@
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Common/interface/TriggerNames.h"
+#include "DQMServices/Core/interface/DQMGlobalEDAnalyzer.h"
+#include "DataFormats/Common/interface/TriggerResults.h"
+#include "DataFormats/HLTReco/interface/TriggerEvent.h"
+
+#include "TrigObjTnPHistColl.h"
+
+
+
+class TrigObjTnPSource : public DQMGlobalEDAnalyzer<std::vector<TrigObjTnPHistColl> >  {
+public:
+  explicit TrigObjTnPSource(const edm::ParameterSet&);
+  ~TrigObjTnPSource() override = default;
+  TrigObjTnPSource(const TrigObjTnPSource&) = delete; 
+  TrigObjTnPSource& operator=(const TrigObjTnPSource&) = delete; 
+  
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  void dqmAnalyze(const edm::Event&, const edm::EventSetup&, const std::vector<TrigObjTnPHistColl>&)const override;
+  void bookHistograms(DQMStore::ConcurrentBooker&, const edm::Run&, const edm::EventSetup& ,std::vector<TrigObjTnPHistColl>&) const override;
+  void dqmBeginRun(const edm::Run&, const edm::EventSetup&, std::vector<TrigObjTnPHistColl>&) const override;
+
+private:
+  edm::EDGetTokenT<trigger::TriggerEvent> trigEvtToken_;
+  edm::EDGetTokenT<edm::TriggerResults> trigResultsToken_;
+  std::string hltProcess_;
+  std::vector<edm::ParameterSet> histCollConfigs_;
+ 
+};
+
+TrigObjTnPSource::TrigObjTnPSource(const edm::ParameterSet& config):
+  trigEvtToken_(consumes<trigger::TriggerEvent>(config.getParameter<edm::InputTag>("triggerEvent"))),
+  trigResultsToken_(consumes<edm::TriggerResults>(config.getParameter<edm::InputTag>("triggerResults"))),
+  hltProcess_(config.getParameter<edm::InputTag>("triggerResults").process()),
+  histCollConfigs_(config.getParameter<std::vector<edm::ParameterSet>>("histColls"))
+{
+    
+}
+
+
+void TrigObjTnPSource::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("triggerEvent", edm::InputTag("hltTriggerSummaryAOD::HLT"));
+  desc.add<edm::InputTag>("triggerResults", edm::InputTag("TriggerResults::HLT"));
+  desc.addVPSet("histColls",
+		TrigObjTnPHistColl::makePSetDescription(),
+		std::vector<edm::ParameterSet>());  
+  descriptions.add("trigObjTnPSource",desc);
+}
+
+void TrigObjTnPSource::bookHistograms(DQMStore::ConcurrentBooker& iBooker,const edm::Run& run,
+				      const edm::EventSetup& setup,
+				      std::vector<TrigObjTnPHistColl>& tnpHistColls)const
+{
+  tnpHistColls.clear();
+  HLTConfigProvider hltConfig;
+  bool hltChanged=false;
+  hltConfig.init(run, setup, hltProcess_, hltChanged);
+  for(auto& histCollConfig : histCollConfigs_) tnpHistColls.emplace_back(TrigObjTnPHistColl(histCollConfig));
```

Comment:
Would it make sense to call this part in the constructor, to avoid storing a copy of the `ParameterSet`s ?

Reply:
hmm are you saying create the TrigObjTnPHistColls in the constructor of the of the source module. Then for each run, copy these histograms into the vector for the run and then initialise. Damn that would have solved my GenericTriggerEventFlag issue, didnt think of that. 

We either have to keep around a vector of TrigObjTnPHistColls or a config to make them so we're wasting space either way.  Given there is no current reason to choose either way, I would prefer to keep the configs and make them on demand as I think its a bit more clear in intention than creating a set of TrigObjTnPHistColls that should never be filled or booked.  However it does mean config errors are found at new run rather than construction. Will think about it but will likely leave as is unless you have strong opinions